src/groupeddataframe/groupeddataframe.jl

```diff
@@ -521,6 +521,131 @@ end
 
 Base.getproperty(key::GroupKey, p::AbstractString) = getproperty(key, Symbol(p))
 
+Base.hash(key::GroupKey, h::UInt) = _nt_like_hash(key, h)
+
+function Base.:(==)(k1::GroupKey, k2::GroupKey)
+    parent(k1).cols == parent(k2).cols || return false
+    return all(((a, b),) -> a == b, zip(k1, k2))
+end
+
+function Base.:(==)(k1::GroupKey, k2::DataFrameRow)
+    parent(k1).cols == _names(k2) || return false
+    return all(((a, b),) -> a == b, zip(k1, k2))
+end
+
+Base.:(==)(k1::DataFrameRow, k2::GroupKey) = ==(k2, k1)
+
+function Base.:(==)(k1::GroupKey, k2::NamedTuple)
+    parent(k1).cols == _propertynames(k2) || return false
+    return all(((a, b),) -> a == b, zip(k1, k2))
+end
+
+Base.:(==)(k1::NamedTuple, k2::GroupKey) = ==(k2, k1)
+
+function Base.isequal(k1::GroupKey, k2::GroupKey)
+    parent(k1).cols == parent(k2).cols || return false
+    return all(((a, b),) -> isequal(a, b), zip(k1, k2))
+end
+
+function Base.isequal(k1::GroupKey, k2::DataFrameRow)
+    parent(k1).cols == _names(k2) || return false
+    return all(((a, b),) -> isequal(a, b), zip(k1, k2))
+end
+
+Base.isequal(k1::DataFrameRow, k2::GroupKey) = ==(k2, k1)
+
+function Base.isequal(k1::GroupKey, k2::NamedTuple)
+    parent(k1).cols == _propertynames(k2) || return false
+    return all(((a, b),) -> isequal(a, b), zip(k1, k2))
+end
+
+Base.isequal(k1::NamedTuple, k2::GroupKey) = ==(k2, k1)
```

Comment:
Maybe this duplication could be avoided by using an `@eval` loop? The only difference is the function which is used with `all`.

Reply:
I will remove most of these codes anyway (and then I will look at how to reduce duplication).
However I am not super fan of `@eval` as then the stacktrace users receive when they make some error are very hard to understand for non-experts.